### PR TITLE
Centrifuge App: Fix finance and repayment forms

### PR DESCRIPTION
**Pull request by sophialittlejohn**
### Description

This pull request...
- updates all currency inputs with thousand separator
    - invest form (with max button)
    - redeem form (with max button)
    - finance form (with max button)
    - repay form (with max button)
    - create pool --> initial max reserve field
    - price asset --> collateral value field
    - sell asset --> price field
- rounds max button input down to 2 decimal places
- adds decimals to invest/redeem/finance/repayment forms to avoid losing precision on max button press
- fixes some broken (back)links
- removes the `=` in the pricing form (#874)
- implements change requests in #846 and #844
- allows financing after the asset has been fully financed and repaid (#886) 

#846 
#844
#874
#886 

### Approvals

- [ ] Dev

### CODE REVIEW

File: centrifuge-app/src/utils/formatting.ts [hunk 1]
```diff
@@ -48,3 +48,7 @@ export function formatPercentage(amount: Perquintill | Decimal | number, include
   })
   return includeSymbol ? `${formattedAmount}%` : formattedAmount
 }
+
+export function roundDown(float: Decimal | number) {
+  return Math.floor((float instanceof Decimal ? float.toNumber() : float) * 100) / 100
+}
```
Comment (offerijns) [inline centrifuge-app/src/utils/formatting.ts:54]: This should probably accept a precision. Now it's implicitly rounding to a precision of 2 decimals. Something like:

```suggestion
export function roundDown(float: Decimal | number, precision: number = 2) {
  return Math.floor((float instanceof Decimal ? float.toNumber() : float) * 10**precision) / 10**precision
}
```

File: fabric/src/components/CurrencyInput/index.tsx [hunk 2]
```diff
@@ -9,9 +9,12 @@ export type CurrencyInputProps = React.InputHTMLAttributes<HTMLInputElement> &
   Omit<InputBoxProps, 'inputElement' | 'rightElement'> & {
     currency?: string
     onSetMax?: () => void
+    variant?: 'small' | 'large'
+    handleChange?: (value: number) => void
```
Comment (offerijns) [inline fabric/src/components/CurrencyInput/index.tsx:13]: I think `onChange` is more standard, and in line with `onSetMax`?

```suggestion
    onChange?: (value: number) => void
```

File: fabric/src/components/CurrencyInput/index.tsx [hunk 3]
```diff
@@ -68,8 +91,26 @@ export const CurrencyInput: React.FC<CurrencyInputProps> = ({
...
+    const inputFormatted = formatThousandSeparator(value)
+    const inputAsNumber = parseFloat(value.replaceAll(',', ''))
+    if (inputProps?.handleChange) {
+      inputProps?.handleChange(inputAsNumber)
+      setValue(inputFormatted)
+    }
+  }
+
+  React.useEffect(() => {
+    const inputFormatted = formatThousandSeparator(Math.floor((inputProps.value as number) * 100) / 100)
```
Comment (offerijns) [inline fabric/src/components/CurrencyInput/index.tsx:110]: Similar to `roundDown`, the decimals shown should be configurable, using a `decimals` argument on `CurrencyInput`.

**Review by offerijns — CHANGES_REQUESTED** [review]
Looks great! Tested the thousand separators in the preview, works perfectly 🔥 Just left a few minor comments.

File: fabric/src/components/CurrencyInput/index.tsx (hunk 3 above)
Comment (sophialittlejohn) [reply fabric/src/components/CurrencyInput/index.tsx]: good call, we would definitely need it sooner or later 👍  I defaulted the precisions to `2` everywhere

File: centrifuge-app/src/pages/Loan/FinanceForm.tsx [hunk 4]
```diff
@@ -67,6 +69,16 @@ export const FinanceForm: React.VFC<{ loan: LoanType }> = ({ loan }) => {
   const maxRepay = balance.lessThan(loan.outstandingDebt.toDecimal()) ? balance : loan.outstandingDebt.toDecimal()
   const canRepayAll = debtWithMargin.lte(balance)
 
+  const allowedToBorrow: Record<LoanInfo['type'], boolean> = {
```
Comment (offerijns) [inline centrifuge-app/src/pages/Loan/FinanceForm.tsx:72]: I'm not sure I follow this, why is this always `false` for bullet loans and credit lines?

Comment (sophialittlejohn) [reply centrifuge-app/src/pages/Loan/FinanceForm.tsx:72]: I added validation for BulletLoans and CreditLines. Can you please double check the logic? It seems a little too simple the way it is now...

File: centrifuge-app/src/pages/Loan/FinanceForm.tsx [hunk 5]
```diff
@@ -65,8 +66,21 @@ export const FinanceForm: React.VFC<{ loan: LoanType }> = ({ loan }) => {
...
+    CreditLineWithMaturity:
+      'maturityDate' in loan.loanInfo &&
+      Dec(daysBetween(loan.originationDate, loan.loanInfo.maturityDate)).gt(0) &&
+      loan.outstandingDebt.toDecimal().lt(availableFinancing),
+    BulletLoan:
+      'maturityDate' in loan.loanInfo &&
+      Dec(daysBetween(loan.originationDate, loan.loanInfo.maturityDate)).gt(0) &&
+      loan.totalRepaid.toDecimal().lt(initialCeiling),
```
Comment (offerijns) [inline centrifuge-app/src/pages/Loan/FinanceForm.tsx:80]: I think this should be:
```suggestion
      loan.totalBorrowed.toDecimal().lt(initialCeiling),
```

**Review by offerijns — CHANGES_REQUESTED** [review]

File: centrifuge-app/src/pages/Loan/FinanceForm.tsx (hunk 5 above)
Comment (sophialittlejohn) [reply centrifuge-app/src/pages/Loan/FinanceForm.tsx:80]: good point, just noticed that with `totalRepaid` the form would still be open even if the `maxBorrowed` is reached

**Review by offerijns — APPROVED** [review]
LGTM 👍

**Pull request merged**